File: backend/app/services/normalizer.py
import hashlib
import re
import uuid
from datetime import datetime, timezone
from decimal import Decimal, InvalidOperation
from typing import Any, Optional
from dateutil import parser as date_parser
from app.utils.timezone import ensure_utc


class Normalizer:
    """Financial data normalization engine."""

    @staticmethod
    def normalize_date(date_raw: Any) -> datetime:
        """
        Parse and normalize various date string formats into timezone-aware UTC datetime.
        Supports ISO, US (MM/DD/YYYY), UK/EU (DD/MM/YYYY), and human date formats.
        """
        if isinstance(date_raw, datetime):
            return ensure_utc(date_raw)

        if not date_raw or not isinstance(date_raw, str):
            raise ValueError(f"Invalid date value: {date_raw}")

        date_str = date_raw.strip()
        if not date_str:
            raise ValueError("Empty date string.")

        try:
            # First attempt standard ISO / dateutil parsing
            dt = date_parser.parse(date_str)
            return ensure_utc(dt)
        except Exception:
            pass

        # Fallback explicit pattern parsing
        patterns = [
            "%Y-%m-%d",
            "%Y/%m/%d",
            "%d/%m/%Y",
            "%m/%d/%Y",
            "%d-%m-%Y",
            "%m-%d-%Y",
            "%d %b %Y",
            "%d %B %Y",
            "%b %d, %Y",
            "%B %d, %Y",
            "%Y-%m-%d %H:%M:%S",
            "%d/%m/%Y %H:%M:%S",
            "%m/%d/%Y %H:%M:%S",
        ]
        for pattern in patterns:
            try:
                dt = datetime.strptime(date_str, pattern)
                return ensure_utc(dt)
            except ValueError:
                continue

        raise ValueError(f"Unable to parse date format: '{date_str}'")

    @staticmethod
    def normalize_amount(
        amount_raw: Any,
        debit_raw: Any = None,
        credit_raw: Any = None,
    ) -> Decimal:
        """
        Parse raw financial amount representations into high-precision Decimal.
        Handles currencies, thousand commas, parentheses for negatives '(500.00)', and separate Debit/Credit columns.
        """
        # If separate Debit / Credit columns exist
        if debit_raw is not None or credit_raw is not None:
            debit_val = Normalizer._parse_single_amount(debit_raw) if debit_raw else Decimal("0.0000")
            credit_val = Normalizer._parse_single_amount(credit_raw) if credit_raw else Decimal("0.0000")
            # Net amount: credit (incoming +) minus debit (outgoing -)
            if credit_val != 0:
                return credit_val.quantize(Decimal("0.0001"))
            elif debit_val != 0:
                # If debit is positive, negate it to reflect cash outflow
                return (-abs(debit_val)).quantize(Decimal("0.0001"))

        return Normalizer._parse_single_amount(amount_raw)

    @staticmethod
    def _parse_single_amount(val: Any) -> Decimal:
        if isinstance(val, Decimal):
            return val.quantize(Decimal("0.0001"))
        if isinstance(val, (int, float)):
            return Decimal(str(val)).quantize(Decimal("0.0001"))

        if not val or not isinstance(val, str):
            raise ValueError(f"Invalid monetary value: {val}")

        clean = val.strip()
        if not clean:
            raise ValueError("Empty monetary amount.")

        # Check for parentheses negative representation e.g. (1,250.50)
        is_negative = False
        if clean.startswith("(") and clean.endswith(")"):
            is_negative = True
            clean = clean[1:-1].strip()
        elif clean.endswith("-"):
            is_negative = True
            clean = clean[:-1].strip()
        elif clean.startswith("-"):
            is_negative = True
            clean = clean[1:].strip()

        # Remove currency symbols ($ € £ ¥ ₹ USD EUR GBP etc.) and thousand separators (commas, spaces)
        clean = re.sub(r"[^\d.]", "", clean)

        if not clean:
            raise ValueError(f"Cannot extract numeric amount from: '{val}'")

        try:
            amount_decimal = Decimal(clean)
            if is_negative:
                amount_decimal = -amount_decimal
            return amount_decimal.quantize(Decimal("0.0001"))
        except (InvalidOperation, ValueError) as exc:
            raise ValueError(f"Invalid decimal amount '{val}': {exc}")

    @staticmethod
    def normalize_description(raw_description: Optional[str]) -> str:
        """
        Normalize description for fuzzy and deterministic matching:
        - Trim leading/trailing whitespace
        - Normalize multiple whitespace and tabs to single spaces
        - Lowercase
        - Strip noise characters
        """
        if not raw_description:
            return ""

        # Normalize whitespace
        desc = re.sub(r"\s+", " ", raw_description.strip())
        # Lowercase
        desc = desc.lower()
        # Remove noisy punctuation while retaining alphanumeric and spaces
        desc = re.sub(r"[^\w\s-]", "", desc)
        return desc.strip()

    @staticmethod
    def normalize_reference(raw_reference: Optional[str]) -> Optional[str]:
        """Normalize transaction external reference codes."""
        if not raw_reference or not isinstance(raw_reference, str):
            return None

        clean = raw_reference.strip()
        if not clean or clean.upper() in {"N/A", "NONE", "NULL", "-", "0"}:
            return None

        # Clean whitespace and standardize uppercase
        clean = re.sub(r"\s+", "", clean).upper()
        return clean

    @staticmethod
    def normalize_currency(raw_currency: Optional[str]) -> str:
        """Validate and standardize 3-letter ISO currency code."""
        if not raw_currency or not isinstance(raw_currency, str):
            return "USD"

        clean = raw_currency.strip().upper()
        # Ensure 3-letter alpha currency code
        if re.match(r"^[A-Z]{3}$", clean):
            return clean
        return "USD"

    @staticmethod
    def compute_transaction_hash(
        source: str,
        organization_id: uuid.UUID,
        transaction_date: datetime,
        amount: Decimal,
        currency: str,
        normalized_description: str,
        normalized_reference: Optional[str] = None,
    ) -> str:
        """
        Compute deterministic SHA-256 fingerprint for deduplication and matching assistance.
        """
        date_str = transaction_date.strftime("%Y-%m-%d")
        amount_str = f"{amount:.4f}"
        ref_str = normalized_reference or ""
        canonical_str = f"{source.lower()}:{str(organization_id)}:{date_str}:{amount_str}:{currency.upper()}:{normalized_description}:{ref_str}"
        return hashlib.sha256(canonical_str.encode("utf-8")).hexdigest()

File: server/src/utils/normalizer.ts
import crypto from "crypto";

export class Normalizer {
  /**
   * Parse various date strings into standard UTC Date objects.
   */
  static normalizeDate(rawDate: string | Date | null | undefined): Date {
    if (!rawDate) throw new Error("Empty date string provided.");
    if (rawDate instanceof Date) return isNaN(rawDate.getTime()) ? new Date() : rawDate;

    const trimmed = String(rawDate).trim();
    if (!trimmed) throw new Error("Empty date string provided.");

    // YYYY-MM-DD
    const isoMatch = /^(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})(?:T.*)?$/.exec(trimmed);
    if (isoMatch) {
      const year = parseInt(isoMatch[1], 10);
      const month = parseInt(isoMatch[2], 10) - 1;
      const day = parseInt(isoMatch[3], 10);
      return new Date(Date.UTC(year, month, day));
    }

    // MM/DD/YYYY or DD/MM/YYYY
    const slashMatch = /^(\d{1,2})[/.-](\d{1,2})[/.-](\d{4})$/.exec(trimmed);
    if (slashMatch) {
      const p1 = parseInt(slashMatch[1], 10);
      const p2 = parseInt(slashMatch[2], 10);
      const year = parseInt(slashMatch[3], 10);

      // If p1 > 12, it must be DD/MM/YYYY
      if (p1 > 12) {
        return new Date(Date.UTC(year, p2 - 1, p1));
      }
      // Otherwise default to MM/DD/YYYY
      return new Date(Date.UTC(year, p1 - 1, p2));
    }

    // Direct ISO or standard parse fallback
    const parsed = new Date(trimmed);
    if (!isNaN(parsed.getTime())) {
      return parsed;
    }

    throw new Error(`Unable to parse date format: '${trimmed}'`);
  }

  /**
   * Parse amounts supporting currency symbols, thousand commas, negative parentheses, and Debit/Credit columns.
   */
  static normalizeAmount(
    rawAmount?: string | number | null,
    rawDebit?: string | number | null,
    rawCredit?: string | number | null,
  ): number {
    // 1. Separate Debit & Credit columns
    if (rawDebit !== undefined && rawDebit !== null && String(rawDebit).trim() !== "") {
      const dr = this.cleanNumericString(String(rawDebit));
      return -Math.abs(dr);
    }
    if (rawCredit !== undefined && rawCredit !== null && String(rawCredit).trim() !== "") {
      const cr = this.cleanNumericString(String(rawCredit));
      return Math.abs(cr);
    }

    if (rawAmount === undefined || rawAmount === null) {
      throw new Error("Empty monetary amount provided.");
    }

    const str = String(rawAmount).trim();
    if (!str) throw new Error("Empty monetary amount string.");

    return this.cleanNumericString(str);
  }

  private static cleanNumericString(val: string): number {
    let clean = val.trim();
    let isNegative = false;

    // Handle parentheses negatives: (500.00) -> -500.00
    if (clean.startsWith("(") && clean.endsWith(")")) {
      isNegative = true;
      clean = clean.slice(1, -1).trim();
    } else if (clean.endsWith("-")) {
      isNegative = true;
      clean = clean.slice(0, -1).trim();
    } else if (clean.startsWith("-")) {
      isNegative = true;
      clean = clean.slice(1).trim();
    }

    // Remove currency symbols ($, €, £, ₹), thousand commas, and spaces
    clean = clean.replace(/[$€£₹,\s]/g, "");

    const num = parseFloat(clean);
    if (isNaN(num)) {
      throw new Error(`Cannot extract numeric amount from '${val}'`);
    }

    const finalNum = isNegative ? -Math.abs(num) : num;
    return Math.round(finalNum * 10000) / 10000;
  }

  /**
   * Normalize description for deterministic and fuzzy matching.
   */
  static normalizeDescription(rawDescription?: string | null): string {
    if (!rawDescription) return "";

    let desc = String(rawDescription).trim().toLowerCase();
    // Normalize multiple spaces/tabs to single space
    desc = desc.replace(/\s+/g, " ");
    // Remove noisy punctuation while retaining alphanumeric and hyphens
    desc = desc.replace(/[^\w\s-]/g, "");
    return desc.trim();
  }

  /**
   * Normalize external reference codes.
   */
  static normalizeReference(rawRef?: string | null): string | null {
    if (!rawRef) return null;
    const clean = String(rawRef).trim().toUpperCase();
    if (!clean || ["N/A", "NONE", "NULL", "-", "0"].includes(clean)) {
      return null;
    }
    return clean.replace(/\s+/g, "");
  }

  /**
   * Deterministic SHA-256 hash for deduplication and integrity check.
   */
  static computeTransactionHash(
    source: string,
    orgId: string,
    date: Date,
    amount: number,
    currency: string,
    rawDesc: string,
    rawRef?: string | null,
  ): string {
    const payload = `${source}|${orgId}|${date.toISOString()}|${amount.toFixed(4)}|${currency.toUpperCase()}|${this.normalizeDescription(rawDesc)}|${this.normalizeReference(rawRef) || ""}`;
    return crypto.createHash("sha256").update(payload).digest("hex");
  }
}
